akka-guard-core-typed/src/main/scala/com/chatwork/akka/guard/typed/SABActor.scala

```diff
@@ -0,0 +1,328 @@
+      case msg: Message =>
+        val future =
+          try {
+            msg.execute
+          } catch {
+            case NonFatal(cause) => Future.failed(cause)
+          }
+        future.onComplete {
+          case Failure(_)                 => context.self ! Failed(failureCount)
+          case Success(r) if isFailed(r)  => context.self ! Failed(failureCount)
+          case Success(r) if !isFailed(r) => context.self ! BecameClosed(attempt, 0, setTimer = false)
+        }(context.executionContext)
+        reply(future, msg.replyTo)
+    }
+
+    private def commonWithClosed(attempt: Long, failureCount: Long): Behavior[Command] = {
+      Behaviors
+        .receiveMessagePartial(common(attempt) orElse closed(attempt, failureCount))
+        .receiveSignal {
+          case (_, PostStop) => postStop()
+        }
+    }
+
+    def behavior: Behavior[Command] = {
+      preStart()
+      commonWithClosed(0, 0)
+    }
+  }
+
+  class ExponentialBackoffActor[T, R](
+      id: ID,
+      maxFailures: Long,
+      backoff: ExponentialBackoff,
+      failureTimeout: FiniteDuration,
+      failedResponse: => Try[R],
+      isFailed: R => Boolean,
+      eventHandler: Option[(ID, SABStatus) => Unit]
+  )(implicit
+      context: ActorContext[Command],
+      timers: TimerScheduler[Command]
+  ) extends SABActor[T, R](id, maxFailures, failureTimeout, failedResponse, isFailed, eventHandler) {
+
+    protected def createScheduler(delay: FiniteDuration, attempt: Long)(key: Any): Unit =
+      timers.startSingleTimer(key, BecameClosed(attempt, 0, setTimer = true), delay)
+
+    private def createResetBackoffSchedule(key: Any): Unit = {
+      backoff.backoffReset match {
+        case AutoReset(resetBackoff) =>
+          createScheduler(resetBackoff, 0)(key)
+        case _ =>
+          ()
+      }
+    }
+
+    override protected def createResetBackoffSchedule(attempt: Long)(key: Any): Unit = {
+      val d = backoff.toDuration(attempt)
+      if (backoff.maxBackoff <= d)
+        createResetBackoffSchedule(key)
+      else
+        createScheduler(d, attempt)(key)
+    }
+
+    override protected def reset(attempt: Long): Behavior[Command] = {
+      createFailureTimeoutSchedule()
+      becomeClosed(attempt, failureCount = 0)
+    }
+  }
+
+  object ExponentialBackoffActor {
+
+    def apply[T, R](
+        id: ID,
+        maxFailures: Long,
+        backoff: ExponentialBackoff,
+        failureTimeout: FiniteDuration,
+        failedResponse: => Try[R],
+        isFailed: R => Boolean,
+        eventHandler: Option[(ID, SABStatus) => Unit]
+    ): Behavior[Command] =
+      Behaviors.setup { implicit context =>
+        Behaviors.withTimers { implicit timers =>
+          val actor = new ExponentialBackoffActor[T, R](
+            id,
+            maxFailures,
+            backoff,
+            failureTimeout,
+            failedResponse,
+            isFailed,
+            eventHandler
+          )
+          actor.behavior
```

Comment:
I think it would be better to implement it using the AbstractBehavior style.

Reply:
@j5ik2o AbstractBehavior seems a bit painful, so I would like to continue with this.